### PR TITLE
Clear Scratch Diffs to Prevent Contaminating Backward through Splits

### LINKED ISSUE #2895: Incorrect gradient from a SoftmaxWithLossLayer with loss_weight 0
I was debugging a network with two loss layers and wanted to disable one of them (a SoftmaxWithLossLayer), as such I set the loss_weight to 0. However, this does not do what I expected at all. The clearest way to explain this is probably using an example on how to reproduce it.

To reproduce one can take the examples/mnist/lenet_train_test.prototxt and add a second loss layer, with weight 0:

```
layer {
 name: "bad_loss"
 type: "SoftmaxWithLoss"
 bottom: "ip2"
 bottom: "label"
 top: "bad_loss"
 loss_weight: 0
}
```

and then run this python script:

``` python
caffe_root = '/roaming/nanne/caffe/' # Update this path to the correct path 
import sys
sys.path.insert(0, caffe_root + 'python')
import os
os.chdir(caffe_root)
import caffe
import numpy as np

caffe.set_mode_gpu()
solver = caffe.SGDSolver(caffe_root + 'examples/mnist/lenet_solver.prototxt')

solver.step(1)

print solver.net.blobs['ip2_ip2_0_split_0'].diff.squeeze()[5:7, :]
print solver.net.blobs['ip2_ip2_0_split_1'].diff.squeeze()[5:7, :]

print solver.net.blobs['ip2'].diff.squeeze()[5:7, :]
```

The diff for the split belonging to the SoftmaxWithLoss with loss_weight 0 will contain 64 (batchsize) values equal to the loss (NOT the gradient) for that input, and all the other elements will be 0. The other split will correctly contain all the diff values (64*10) for the loss with weight 1.

However, these two splits still get combined, creating the diff for 'ip2' for which the first 64 values are not comparable to the last 576. Am I wrong in how I tried to use the loss_weight or is this a bug? (It doesn't seem to be specific to SoftmaxWithLoss, though its most clear for this layer).

### LINKED ISSUE #5981: New Accuracy Layer on GPU interferes with training
### Issue summary
Using the "Accuracy" layer in the "Training net" on GPU breaks the training. The layer somehow interferes with the gradient. Loss explodes quickly and Train/Test Accuracies stall to 1.

### Steps to reproduce
1. Download the the latest version of Caffe (commit 691febcb83d6a3147be8e9583c77aefaac9945f8)
2. Compile it with this [Makefile.config](https://github.com/BVLC/caffe/files/1385875/Makefile.txt)
3. Run a network with Accuracy layer in the TRAIN phase.

### My system configuration
Operating system: Ubuntu 14.04.5 LTS
CUDA version (if applicable): 7.0
CUDNN version (if applicable): 4.7
BLAS: libblas.so.3
Python version: 3.5

### How to fix it
All these three solutions can fix it:
- Remove the Accuracy Layer from the "Training net", not problems with the phase: TEST.
- Change back-end to CPU.
- Rollback before commit 62e0c8559045cb2b5a12e0d6c41acd25d412263 (which I suspect caused the issue).

### PR DESCRIPTION
**Pull request by shelhamer**
Certain layers (`SoftmaxWithLossLayer`, `SigmoidCrossEntropyLossLayer`, and `AccuracyLayer`) save memory during forward by making use of bottom diffs that are otherwise unused or overwritten during backward. The trouble is that these scratch diffs can be mistakenly propagated by backward through split layers. All of the top diffs of a split are accumulated even when backward is not called for the losses (when their loss weight is zero) or accuracy (which has no backward step). This was missed at first because it requires the interaction of split layers and the backward pruning that prevents computation of unnecessary gradients.

This fix zeros out the scratch diffs to prevent this kind of contamination. This requires a little computation but not much at all. This is preferable to requiring more memory for internal buffers because the further memory usage might cause an unexpected crash for a previously good configuration.

See https://github.com/BVLC/caffe/issues/2895#issuecomment-130984651 for the first explanation of this issue.

Related:
- fix #2895, fix #5981
- close #3868, close #5987, close #6141

Credits:
- Thanks @vlomonaco for reporting the `Accuracy` layer issue in #5981.
- Thanks @Noiredd for proposing an internal buffer fix in #5987 and @sonack for another version of the same in #6141
- Thanks @orzlibo for proposing the clearing fix for losses in #6186. Note the
cherry-pick of that fix in this PR.

### CODE REVIEW

**Comment by Noiredd** [comment]
This is a neat, systemic solution. Performance impact of `caffe_gpu_set` is minimal: on a theoretical network making heavy use of Accuracy it doubled the forward pass time... from 0.78ms to 1.60ms - IMO negligible. The same network ate 703 MB RAM with my hotfix, comparable to 526 MB with this PR. On more typical classification setups the differences are unnoticeable.  
My vote goes for merging this.

I have added a reference to close #6141 which attempted to solve the Accuracy layer problem by allocating a raw SyncedMemory object to act as an internal buffer. Thanks @sonack!

**Comment by shelhamer** [comment]
Thanks all for your work on this and thanks again @Noiredd for the final review!
